### PR TITLE
[Feed Reader] Data layer

### PR DESCRIPTION
**Pull request by aurimasv**
Let's try this.

I still need to write more tests (and there are quite possibly some additional issues with the DB transactions), so don't read too much into it. It would probably be best to just look at the existing tests for now.

### CODE REVIEW

File: test/tests/feedTest.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+			assert.isUndefined(feed.lastUpdate, "lastUpdate is null");
+			assert.isUndefined(feed.lastCheckError, "lastCheckError is null");
+		});
+		it("should not allow setting a feed as child of another feed", function *() {
+			let feed = new Zotero.Feed();
+			
+			assert.throws(() => feed.parentID = 1, 'Cannot set parent key for feed', 'cannot set parent by ID');
+			assert.throws(() => feed.parentKey = 1, 'Cannot set parent key for feed', 'cannot set parent by key');
+		});
+		it("should not allow saving a feed with the same url", function *() {
```
Comment (aurimasv) [inline test/tests/feedTest.js]: I'm not 100% set on this one, but I also don't see any reason to allow it. Interestingly, the test actually currently fails, even though the SQLite schema specifies UNIQUE for url.

File: test/tests/feedTest.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	describe("#erase()", function() {
+		it("should erase a saved feed", function *() {
+			let feed = new Zotero.Feed();
+			feed.name = 'Foo';
+			feed.url = 'http://www.exampled.com';
+			let id = yield feed.saveTx({skipEditCheck: true});
+			feed = yield Zotero.Feeds.getAsync(id);
+			yield feed.eraseTx();
+			
+			yield assert.isRejected(feed.eraseTx(), "does not allow erasing feed twice");
```
Comment (aurimasv) [inline test/tests/feedTest.js]: This currently does not thorw/rejects and I guess it's not a big deal, but IMO it should be an error to try and delete an object twice.

File: test/tests/feedTest.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			
+			feed = yield Zotero.Feeds.getAsync(id);
+			assert.notOk(feed, "erased feed no longer exists");
+		});
+		it("should throw if trying to erase an unsaved feed", function() {
```
Comment (aurimasv) [inline test/tests/feedTest.js]: Same for this

File: test/tests/librariesTest.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+		assert.isAbove(Zotero.Libraries.feedLibraryID, -1, "#feedLibraryID is a positive integer");
+	});
+	describe("#getAll()", function() {
+		let ids = Zotero.Libraries.getAll();
+		
+		it("should return an array of valid library IDs", function() {
+			assert.isArray(ids, "returns an array");
+			assert(ids.reduce(function(res, id) { return res && Number.isInteger(id) && id >= 0 }, true), "All IDs are positive integers");
+		})
+		it("should return all library IDs", function *() {
```
Comment (aurimasv) [inline test/tests/librariesTest.js]: Interestingly, this currently fails https://travis-ci.org/zotero/zotero/jobs/65026201#L516 but I haven't investigated why yet.

File: test/tests/feedTest.js (hunk 1 above)
Comment (dstillman) [inline test/tests/feedTest.js]: Yeah, I'm fine with restricting this.

**Comment by aurimasv** [comment]
Sorry @dstillman, this PR inflated a bit, but I think it's mostly ok now, so please take a look when you get a chance.

I'll add a couple notes to these commits

**Comment by aurimasv** [comment]
I guess I can make this a bit easier. Please give me a few minutes to clean up the commits.

**Comment by dstillman** [comment]
So I guess this would've been a good question to ask a long time ago, but why implement feeds as collections and not as individual libraries, the same way that groups are just libraries with some extra data in another table?

**Comment by aurimasv** [comment]
What would be the advantage? I think they're conceptually similar, but here are some advantages of a collection-like model:
* Initial plan (and still sort of planned) was to make it possible to group different feeds into collections for organizational purposes.
* You could, in theory, have the same items (by GUID) being pulled into different feeds. Reading one item would mark that item as read in all feeds.
* Marking all feeds as read could be easier if they're in the same library (i.e. right-click library and select mark as read)

**Comment by dstillman** [comment]
> Initial plan (and still sort of planned) was to make it possible to group different feeds into collections for organizational purposes.

Ah, OK. I suppose that's sufficient justification. And you might want a recursive-collections display of the sub-feeds from a parent collection.

File: chrome/content/zotero/xpcom/data/dataObjects.js [hunk 5]
```diff
@@ -74,6 +72,9 @@ Zotero.defineProperty(Zotero.DataObjects.prototype, 'primaryFields', {
 	get: function () Object.keys(this._primaryDataSQLParts)
 }, {lazy: true});
 
+Zotero.defineProperty(Zotero.DataObjects.prototype, 'primaryDataSQLFrom', {
+	get: function() " " + this._primaryDataSQLFrom + " " + this._primaryDataSQLWhere
+}, {lateInit: true});
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/dataObjects.js]: This and the change above are actually no longer relevant. Some hacky code relied on this previously.

File: chrome/content/zotero/xpcom/data/feed.js [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	this._feedCleanupAfter = null;
+	this._feedRefreshInterval = null;
+	this._feedUnreadCount = null;
+	
+	if (url) this.url = url;
+}
+
+Zotero.Feed._super = Zotero.Collection;
+Zotero.Feed.prototype = Object.create(Zotero.Feed._super.prototype);
+Zotero.Feed.constructor = Zotero.Feed;
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/feed.js]: note to self: use `Zotero.extendClass`

File: chrome/content/zotero/xpcom/data/feed.js [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+		throw new Error('Feed URL is not set');
+	} else {
+		// Validate URL
+		let uri,
+			invalidUrlError = "Invalid feed URL " + this.url;
+		try {
+			uri = Components.classes["@mozilla.org/network/io-service;1"]
+				.getService(Components.interfaces.nsIIOService)
+				.newURI(this.url, null, null);
+			this.url = uri.spec;
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/feed.js]: note to self: move this down

**Comment by dstillman** [comment]
On the other hand, one could say the same thing about group libraries — and other sorts of future subscriptions, of which there'll likely be some —  and that what we actually need is the ability to group libraries of different types into folders. They could be combined in the interface or in separate sections, but either way we'd want the container to be a folder, not a collection...

**Comment by dstillman** [comment]
I don't think the GUID issue matters (that could easily be applied across libraries), and the marking-as-read wouldn't be an issue as long as there was a common container. (Latter would need to go somewhere else if you could mix and match libraries of different types in folders, but I'm not sure we'd do that. In any case, just a UI issue).

**Comment by aurimasv** [comment]
It's not just about guid, but items being completely different in different libraries, though that's not a show stopper. I'm honestly not really set on one way or another. Either implementation seems ok to me, but I don't really see an advantage to the library implementation. Maybe you have something specific in mind. Changing current implementation would not be difficult.

**Comment by aurimasv** [comment]
I guess one advantage is that deleting a library would delete all items within it. Also items would be guaranteed to be in a library.

**Comment by dstillman** [comment]
Yeah, my sense is that there'll be more that we'll have to override or special-case than if these were libraries. And I think we may want the ability to organize both these and other library types into folders, which wouldn't be possible if these were collections.

I do see the advantage of having items with the same id in different feeds be the same item, but that also seems pretty rare. And even that would require special-casing, because if you removed an item from all feeds it'd still exist in the feed library if we relied on default behavior.

So if you have a decent solution for the marking-as-read-across libraries issue (which, honestly, is optional — and I don't even know for sure if we can trust ids), I think it'd probably make more sense to go the library route.

**Comment by aurimasv** [comment]
OK Dan, here goes. Feeds as Libraries. Zotero.Library is now a class and Zotero.Group and Zotero.Feed both inherit from it. Zotero.Library mostly follows Zotero.DataObject conventions, but doesn't inherit from it directly.

File: chrome/content/zotero/xpcom/data/collection.js [hunk 8]
```diff
@@ -23,7 +23,7 @@
     ***** END LICENSE BLOCK *****
 */
 
-Zotero.Collection = function() {
+Zotero.Collection = function(name) {
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/collection.js]: I've been trying to make constructors take required parameters, since they always need to be set anyway. libraryID defaults to user library, so it's not technically required.

File: chrome/content/zotero/xpcom/data/collections.js [hunk 9]
```diff
@@ -46,13 +46,23 @@ Zotero.Collections = function() {
 		hasChildCollections: "(SELECT COUNT(*) FROM collections WHERE "
 			+ "parentCollectionID=O.collectionID) != 0 AS hasChildCollections",
 		hasChildItems: "(SELECT COUNT(*) FROM collectionItems WHERE "
-			+ "collectionID=O.collectionID) != 0 AS hasChildItems"
+			+ "collectionID=O.collectionID) != 0 AS hasChildItems",
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/collections.js]: hmm, I looks like this may have gotten left over after a merge conflict. Will clean it up.

File: chrome/content/zotero/xpcom/data/collections.js [hunk 10]
```diff
@@ -46,13 +46,23 @@ Zotero.Collections = function() {
...
 	};
 	
 		
 	this._primaryDataSQLFrom = "FROM collections O "
-			+ "LEFT JOIN collections CP ON (O.parentCollectionID=CP.collectionID)";
+			+ "LEFT JOIN collections CP ON (O.parentCollectionID=CP.collectionID)"
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/collections.js]: Undo this as well

File: chrome/content/zotero/xpcom/data/collections.js [hunk 11]
```diff
@@ -46,13 +46,23 @@ Zotero.Collections = function() {
...
 	
+	this._relationsTable = "collectionRelations";
+	
+	// Choose correct item type when loading from DB row
+	this._getObjectForRow = function(row) {
+		if (row.feedUrl) {
+			return new Zotero.Feed();
+		}
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/collections.js]: Another leftover chunk of code (sorry, making notes for myself)

File: chrome/content/zotero/xpcom/data/dataObject.js [hunk 12]
```diff
@@ -1171,9 +1181,13 @@ Zotero.DataObject.prototype._initErase = function (env) {
 		key: this.key
 	};
 	
+	if (!env.options.skipEditCheck) this.editCheck();
+	
 	if (env.options.skipDeleteLog) {
 		env.notifierData[this.id].skipDeleteLog = true;
 	}
+	
+	return Zotero.Promise.resolve(true);
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/dataObject.js]: Should wrap this in `Zotero.Promise.method` instead

File: chrome/content/zotero/xpcom/schema.js [hunk 13]
```diff
@@ -1469,7 +1469,7 @@ Zotero.Schema = new function(){
 			});
 			yield _updateDBVersion('compatibility', _maxCompatibility);
 			
-			var sql = "INSERT INTO libraries (libraryID, libraryType, editable, filesEditable) "
+			var sql = "INSERT INTO libraries (libraryID, type, editable, filesEditable) "
 				+ "VALUES "
 				+ "(?, 'user', 1, 1), "
 				+ "(4, 'publications', 1, 1)"
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/schema.js]: Why are we specifying library IDs here at all? why not just let DB assign those? (granted that they would pretty much always end up 1 and 2 anyway)

File: chrome/content/zotero/xpcom/uri.js [hunk 14]
```diff
@@ -95,13 +95,14 @@ Zotero.URI = new function () {
...
+			case 'feed':
 				var id = Zotero.Users.getCurrentUserID();
 				if (!id) {
 					throw new Zotero.Error("User id not available", "USER_NOT_AVAILABLE");
 				}
 				
-				if (libraryType == 'publications') {
-					return "users/" + id + "/publications";
+				if (libraryType == 'publications' || libraryType == 'feed') {
+					return "users/" + id + "/" + libraryType;
```
Comment (aurimasv) [inline chrome/content/zotero/xpcom/uri.js]: Need to update this. We no longer have a single feed library.

File: test/tests/groupTest.js [hunk 15]
```diff
@@ -1,13 +1,28 @@
-"use strict";
-
```
Comment (aurimasv) [inline test/tests/groupTest.js]: should revert this

File: test/tests/itemsTest.js [hunk 16]
```diff
@@ -2,6 +2,7 @@ describe("Zotero.Items", function () {
 	var win, collectionsView, zp;
 	
 	before(function* () {
+		this.timeout(10000);
 		win = yield loadZoteroPane();
```
Comment (aurimasv) [inline test/tests/itemsTest.js]: This was taking too long when running items tests directly without running other tests first.

File: test/tests/translateTest.js [hunk 17]
```diff
@@ -44,16 +44,18 @@ function saveItemsThroughTranslator(translatorType, items) {
...
-		"	var json = JSON.parse('"+JSON.stringify(items).replace(/'/g, "\'")+"');\n"+
+		"	var json = JSON.parse('"+JSON.stringify(items).replace(/['\\]/g, "\\$&")+"');\n"+
 		"	for (var i=0; i<json.length; i++) {"+
 		"		var item = new Zotero.Item;\n"+
 		"		for (var field in json[i]) { item[field] = json[i][field]; }\n"+
 		"		item.complete();\n"+
 		"	}\n"+
-		"}"));
+		"}"
+	));
```
Comment (aurimasv) [inline test/tests/translateTest.js]: Undo this change

**Comment by aurimasv** [comment]
And I still need to fix some sync issues (see failing test), though I'm not quite yet sure what's going on there.

File: chrome/content/zotero/xpcom/data/collection.js (hunk 8 above)
Comment (dstillman) [inline chrome/content/zotero/xpcom/data/collection.js]: The problem with this is that it's messier and less clear later if we add a required parameter. And setting the name before the libraryID is awkward. In the absence of named parameters, I think I'd prefer to just use properties. (We do use a required parameter for itemTypeID for Zotero.Item, but that one isn't ever going to change, it does have to be set before most other fields, and seems less weird before libraryID because it's actually defining the kind of item you're creating.)

File: resource/schema/userdata.sql [hunk 18]
```diff
@@ -238,11 +257,11 @@ CREATE INDEX deletedItems_dateDeleted ON deletedItems(dateDeleted);
 
 CREATE TABLE libraries (
     libraryID INTEGER PRIMARY KEY,
-    libraryType TEXT NOT NULL,
-    editable INT NOT NULL,
-    filesEditable INT NOT NULL,
+    type TEXT NOT NULL,
+    editable INT NOT NULL DEFAULT 0,
+    filesEditable INT NOT NULL DEFAULT 0,
```
Comment (dstillman) [inline resource/schema/userdata.sql]: I left defaults out for these on purpose because the code creating libraries really does need to make the decision as to editability. Creating a library isn't a common action, so there's no need for this to be implicit.

File: resource/schema/userdata.sql [hunk 19]
```diff
@@ -205,6 +205,25 @@ CREATE TABLE collectionRelations (
...
+    lastCheckError TEXT,
+    cleanupAfter INT,
+    refreshInterval INT,
+    FOREIGN KEY (libraryID) REFERENCES libraries(libraryID) ON DELETE CASCADE
+);
+
+CREATE TABLE feedItems (
+    itemID INTEGER PRIMARY KEY,
+    guid TEXT NOT NULL UNIQUE,
+    readTimestamp TIMESTAMP,
```
Comment (dstillman) [inline resource/schema/userdata.sql]: `readTime`? (Also, we're not actually using the time, are we? Unless there's a compelling need to show or use that anywhere, I think this can just be `read INT` with a boolean.)

Comment (aurimasv) [inline resource/schema/userdata.sql]: We use the time to decide when to purge read items.

Comment (dstillman) [inline resource/schema/userdata.sql]: Ah, OK.

File: chrome/content/zotero/xpcom/schema.js (hunk 13 above)
Comment (dstillman) [inline chrome/content/zotero/xpcom/schema.js]: It's for consistency with migrateUserDataSchema, where we use 1 and 4 (known unused ids, coming from the global libraryIDs in 4.0) for upgrading users. So, actually, the '?' for 'user' should just be 1 here.

**Comment by dstillman** [comment]
I wouldn't worry about the sync failure for now. We'll need to discuss our general syncing plan for feeds.

File: chrome/content/zotero/xpcom/data/collection.js (hunk 8 above)
Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/collection.js]: I see in support.js we pass down `params` object as argument. Is that what we want to do in general?

**Comment by aurimasv** [comment]
>I wouldn't worry about the sync failure for now. We'll need to discuss our general syncing plan for feeds, and I can help with that part.

The changes in this PR affect syncing of group/user libraries though. I'm more than happy to leave that part to you though.

**Comment by aurimasv** [comment]
I think that should address most comments. Still need to do:
* Add tests for Zotero.URI
* Change constructor parameters
* Make table columns unique?

File: chrome/content/zotero/xpcom/data/collection.js (hunk 8 above)
Comment (dstillman) [inline chrome/content/zotero/xpcom/data/collection.js]: I'm OK with a `params` argument, as an alternative to setting properties after the constructor.

**Comment by dstillman** [comment]
Yeah, I can deal with sync breakage. I have a big sync patch coming on my end anyway.

**Comment by aurimasv** [comment]
Please take a look. We're getting close, I think.

**Comment by aurimasv** [comment]
OK, how about now?

File: chrome/content/zotero/xpcom/data/feedItem.js [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+		env.transactionOptions.onCommit = () => {
+			if (superOnCommit) superOnCommit();
+			this.ObjectsClass._setGUIDMapping(this.guid, env.id);
+		};
+	}
+	
+	return proceed;
+});
+
+Zotero.FeedItem.prototype.forceSave = function(options) {
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/data/feedItem.js]: If this is going to to exist, seems like it should at least be `forceSaveTx()` (and `forceEraseTx()` below).

**Comment by dstillman** [comment]
I might have more notes later, and we'll need to make some adjustments once we figure out the organization stuff, but this is looking good, so you can squash it down as appropriate.

**Comment by dstillman** [comment]
(Other than the few failing (non-sync) tests.)

**Comment by aurimasv** [comment]
>(Should fix the failing (non-sync) tests, though.)

>`Cannot edit feedItem in read-only Zotero library`

Is the result of reverting [this](https://github.com/aurimasv/zotero/commit/e097aba50479eede71d4138ed715fa7ca6368760#commitcomment-13321934), so I think it should currently fail because that throw needs to be fixed.

Also, could you OK the last commit before I squash this all down?

**Comment by dstillman** [comment]
OK, all set.

**Comment by aurimasv** [comment]
Squish squashed

File: chrome/content/zotero/xpcom/data/feedItem.js (hunk 20 above)
Comment (dstillman) [inline chrome/content/zotero/xpcom/data/feedItem.js]: Could you adjust these?

Comment (aurimasv) [inline chrome/content/zotero/xpcom/data/feedItem.js]: Sorry about that. Fixed and squashed.

**Comment by aurimasv** [comment]
:tada:

**Comment by zuphilip** [comment]
Wow, :clap: , cool!
